README.md
[![Build status](https://ci.appveyor.com/api/projects/status/0gif1tjb4lmtoro0?svg=true)](https://ci.appveyor.com/project/MSP-Greg/ruby-loco)

This repo contains code to build ruby trunk from a local repo using MinGW and msys2 packages.

Please see [ChangeLog](https://github.com/MSP-Greg/ruby-loco/blob/master/ChangeLog.md) for info regarding recent updates.

### Requirements, Setup and Configuration

1. An msys2/MinGW installation, with toolchain, gcc-libs, libffi, ncurses, readline, termcap, nd zlib installed.
2. A clone or fork of [oneclick/rubyinstaller2](https://github.com/oneclick/rubyinstaller2)
3. [GitHub for Windows](https://git-for-windows.github.io/).
4. Installation of the ruby-gdbm, ruby-openssl, and ruby-libyaml packages from [release v0.1](https://github.com/MSP-Greg/ruby-loco/releases/tag/v0.1).  Existing
package should be removed if they are installed.
5. A symlink named `ruby` within the `src` directory to your ruby repo.  On my system, the command is (run as admin):
```
mklink /d E:\GitHub\ruby-loco\src\ruby E:\GitHub\ruby
```

You may need to create the src dir first, not sure...

6. Since the default `Git for Windows` install path is in `Program Files`, a symlink with no spaces is needed.  I used:
```
mklink /d E:\GitHub\ruby-loco\git "C:\Program Files\Git"
```
7. The file [pkg_set_env.cmd.sample]() needs to copied/renamed to `pkg_set_env.cmd`, with all the environment variables set for your system.  I set it up to use a separate temp directory for building.  Note the path delimiters used.

#### Important Note - I have my most of my 'normal' apps and windows on my c drive, but most of my repos and development code are on my e drive.  Please make sure to adjust for that.

### Use

Before building, position the ruby repo where you would like it (branch, commit, tag, etc).  Code makes no changes to the repo.  At present, builds from 2.3 forward seem to work.

To build and run all tests, run/click [pkg_build_64.cmd](https://github.com/MSP-Greg/ruby-loco/blob/master/pkg_build_64.cmd)

To run test-all again, run [pkg_test-all_64.cmd](https://github.com/MSP-Greg/ruby-loco/blob/master/pkg_test-all_64.cmd).  It will overwrite the *test-all.log file.

A useable ruby build exists in the pkg dir.  I have used these builds to update my doc site, and installed several ruby extension gems.  See [RubyInstaller2 wiki](https://github.com/oneclick/rubyinstaller2/wiki) for information about extension gem installation.

**Important** Between builds, please make sure to reset the Ruby repo with something similar to:

```
git clean -fdx & git reset --hard HEAD
```

### Notes

This code was started based on some forks/branches of the original [oneclick/rubyinstaller](https://github.com/oneclick/rubyinstaller).

Originally, I was just building trunk.  Recently, I build the ruby_2_3 branch, and it required some changes (1.0.2 openssl package and installation dlls).  I have not (yet) created code to allow easy building of versions before 2.5.

I've got an interest in doc software (see [msp-greg.github.io/](https://msp-greg.github.io/)), so I've tried to keep code in either .cmd or .rb files, as there are a few common (deservedly rightly so) std-lib / gem items that do not doc well.

### Packages

GDBM - I believe the current msys2/mingw gdbm 1.12 package does not work.  I have tried to build 1.11, 1.12, and 1.13, to no avail.  The package used is based on 1.10.

LibYAML - some packages have a libyaml.dll file.  Mine does not, and all tests to pass.

OpenSSL - built to use the Windows CAPI engine (enable-capieng & --api=1.0.0).

### rb file info

| File                | runs in          | Desc                                                     |
| ------------------- | ---------------- | -------------------------------------------------------- |
| prepare_pre.rb      | cmd              | Loads env variables with info from repo                  |
| prepare.rb          | PKBBUILD prepare | Cleans some artifacts from src\ruby dir, applies patches, and adds bundled gems |
| install_dll_info.rb | install_post.rb  | Constants that define included dll's by ruby version     |
| install_post.rb     | PKBBUILD package | Copies dlls, creates manifest xml, modifies .exe files   |
| install_post_ri2.rb | PKBBUILD package | Copies RI2 runtime files to package                      |

### Patches

At present, all patches are located in the patches dir.  The sub dirs, 64 and 32, are used when patches differ for 32 and 64 bit builds.

1. The only patches required for building are `configure.in.patch` and `include_ruby_defines.h.patch`.

2. All other patches are for `test-all`.  Patches beginning with `segv` are intermittently needed, otherwise `test-all` silently stops.

3. If a patch file name begins with an underscore, **it is not applied**.

4. `test-runner.patch` adds several things that output info (ENV, PATH, etc) to the test-all.log file.

5. `lib-rubygems-test_case.patch` has an open PR for it.

6. `test-readline-test_readline.patch` is needed due to some issues with readline testing using temp files.

Patch code needs to be written and re-organized to allow patching based on the ruby version being built.  I haven't decided how best to set that up, given the desire to be able to easily swap patches in and out of builds...

### Thanks to...

Lars Kanis for [RubyInstaller2](https://github.com/oneclick/rubyinstaller2).  Code I've written builds and runs test-all, but all of the runtime code, the code that creates the dll manifest (and modifies the exe files to use it), and hooks extension gem build into msys2 came from his work.  Thanks again.


ChangeLog.md
### 2017-05-10

* Added patch for [`lib/mkmf.rb`](https://github.com/MSP-Greg/ruby-loco/blob/master/patches/lib-mkmf.rb.patch), seems to reliably run make with -j3.  Passes test-all.

* Added two encoding related lines at the end of [`pkg_set_env.cmd`](https://github.com/MSP-Greg/ruby-loco/blob/master/pkg_set_env.cmd.sample).  Seems to improve test-all results if these are both set to 'filesystem' encoding.

* Added patches for [`ruby/test_process.rb`](https://github.com/MSP-Greg/ruby-loco/blob/master/patches/test-ruby-test_process.patch) & [`ruby/test_rubyoptions.rb`](https://github.com/MSP-Greg/ruby-loco/blob/master/patches/test-ruby-test_rubyoptions.patch).  These remove three test-all failures.

### 2017-05-06

* Update PKGBUILD for 2017-05-06_58576 - removed ncurses and after-update, which seemed to be causing duplicate tasks with make all.

### 2017-05-03

* Update PKGBUILD for 2017-05-03 58552

### 2017-05-02

* PKGBUILD
  * Added jobs variable
  * Adjusted make command to run correctly

prepare.rb
# frozen_string_literal: true

require 'rubygems'
require 'rubygems/gem_runner'
require 'rubygems/package'

# See ruby tool/gem-unpack.rb
#
# This method is used by "make extract-gems" to unpack bundled gem files.
#
# Unlike the gem unpack command, it correctly unpacks the gemspec file.
# Corrects problems with spec.files & spec.test_files using things like
# git ls-files or grep

def Gem.unpack(file, dir = nil)
  pkg = Gem::Package.new(file)
  spec = pkg.spec
  target = spec.full_name
  target = File.join(dir, target) if dir
  pkg.extract_files target
  spec_file = File.join(target, "#{spec.name}.gemspec")
  open(spec_file, 'wb') do |f|
    f.print spec.to_ruby
  end
  puts "gem unpacked #{file}"
end

module Prepare  

  @@r_inst_dir = "ruby#{ENV['R_VERS_2']}_#{ ENV['MINGW_INSTALLS'][-2,2]}"
  @@arch = @@r_inst_dir[-2,2]

  class << self  

    def run
      @@r_vers_int = vers_str_2_int(ENV['R_VERS'])
      check_openssl
      copy_spec if @@r_vers_int < 20500
      `attrib -r %REPO_RUBY%\\spec\\*.* /s /d`
      apply_patches
      clean_repo
      add_gems
    end

    private

    # Cleans and refreshes bundled gems
    def add_gems
      b_gem_dir = File.join(__dir__, 'src', 'ruby', 'gems')
      hsh = Hash.new
      
      # get list of bundled gems and delete incorrect items or versions
      Dir.chdir(b_gem_dir) { |d|
        # load hash with key of full name ie, 'did_you_mean-1.1.0', and value
        # of array with [name, vers]
        File.open( 'bundled_gems', 'rb') { |f|
          f.read.each_line { |l|
            ary = l.strip.split(/\s+/)
            hsh[ary.join('-')] = ary
          }
        }

        # need to loop thru existing files in gems and delete any that aren't needed
        gem_files = Dir["*.gem"]
        gem_dirs  = Dir["*/"]
        gem_files.each { |f|
          f_name = f.sub(/\.gem/, '')
          `del #{f}` unless hsh.key?(f_name)
        }
        gem_dirs.each { |d|
          d_name = d.sub(/\//, '')
          `rd /s /q #{d_name}`unless hsh.key?(d_name)
        }
      }

      # check for, and add (if needed) all required gems
      `md bundled_gems` unless Dir.exist?('bundled_gems')
      Dir.chdir(File.join(__dir__, 'bundled_gems') ) { |dir|
        win_b_gem_dir = b_gem_dir.gsub(/\//, '\\')
        hsh.each { |k,v|
          # see if they exist already
          next if File.exist?( File.join(b_gem_dir, "#{k}.gem") ) &&
            Dir.exist?( File.join(b_gem_dir, k) )
          unless File.exist?("#{k}.gem") && Dir.exist?(k)
            # need to get gem
            runner ||= Gem::GemRunner.new
            begin
              puts "gem fetch #{v[0]} -v #{v[1]}"
              runner.run ['fetch' , v[0], '-v',  v[1]]
              Gem.unpack("#{k}.gem")
#              puts "gem unpack #{k}"
#              runner.run ['unpack', "#{k}.gem"]
            rescue
              puts "Issue with bundled gem #{k}"
              next
            end
          end
          `copy #{k}.gem #{win_b_gem_dir}`
          unless Dir.exist? File.join(b_gem_dir, k)
            # puts "md #{win_b_gem_dir}\\#{k}"
            `md #{win_b_gem_dir}\\#{k}` 
          end
          # puts "xcopy /s /q #{k} #{win_b_gem_dir}\\#{k}"
          `xcopy /s /q #{k} #{win_b_gem_dir}\\#{k}`
        }
      }
    end

    # Reads patch dir, applies patches
    def apply_patches
      
      raise ArgumentError, "Argument must 32 or 64" unless @@arch == '32' || @@arch == '64'
      # puts "Dir.pwd #{Dir.pwd}"

      ENV['PATH'] += "#{File::PATH_SEPARATOR}#{File.join(__dir__, 'git', 'cmd')}"

      Dir.chdir('src/ruby') { |dir|
        # collect patches and apply
        patches = Dir["#{__dir__}/patches/{[^_],#{@@arch}/[^_]}*.patch"]
        # 
        patches.concat(vers_patches)
        patches.sort_by! { |p| File.basename(p) }
        patches.each { |p|
          puts "#{'- ' * 25} #{File.basename(p)}" 
          puts `patch -p1 -N --no-backup-if-mismatch -i #{p}`
        }
        puts ''  # just for formatting of prepare.log
      }
    end

    # Checks mingw openssl version.  If needed, swaps to correct version
    def check_openssl
      puts "OpenSSL - checking  package version..."
      arch = (@@arch == '32' ? 'i686' : 'x86_64')
      t = `pacman -Qs mingw-w64-#{arch}-openssl`.strip
      if t && !t.empty?
        vers = vers_str_2_int(t.strip.split(' ')[1])
        vers_old = "#{ENV['MSYS2_DIR_U']}/var/cache/pacman/pkg/mingw-w64-#{arch}-openssl-#{ENV['OPENSSL_102']}-any.pkg.tar.xz"
        vers_new = "#{ENV['MSYS2_DIR_U']}/var/cache/pacman/pkg/mingw-w64-#{arch}-openssl-#{ENV['OPENSSL_110']}-any.pkg.tar.xz"
        if @@r_vers_int >= 20400 && vers < 10100
          # swap to 1.1.0
          puts `pacman -Rdd --noconfirm mingw-w64-#{arch}-openssl`
          puts `pacman -Udd --noconfirm --force #{vers_new}`
        elsif @@r_vers_int < 20400 && vers >= 10100
          # swap to 1.0.2
          puts `pacman -Rdd --noconfirm mingw-w64-#{arch}-openssl`
          puts `pacman -Udd --noconfirm --force #{vers_old}`
        end
      else
        if @@r_vers_int >= 20400
          puts `pacman -Udd --noconfirm --force #{vers_new}`
        else
          puts `pacman -Udd --noconfirm --force #{vers_old}`
        end
      end
      puts "OpenSSL - done"
    end

    # Cleans previous build artifacts from ruby repo, all are accounted for in
    # .gitignore.
    #
    def clean_repo
      Dir.chdir( 'src/ruby') { |dir|
#        ruby_files   = %w[ configure tool\\config.guess tool\\config.sub ]
        ruby_files   = %w[ configure ]
        
        ruby_folders = %w[ autom4te.cache ]   # enc\\unicode\\data
        
        ruby_files.each   { |f|
          if File.exist?(f)
            puts "Deleting file   #{f}"
            `del /q #{f}`
          end
        }
        ruby_folders.each { |f|
          if Dir.exist?(f)
            puts "Deleting folder #{f}"
            `rd /s /q #{f}`
          end
        }
      }
      if Dir.exist?("src/build-#{@@arch}")
        `rd /s /q src\\build-#{@@arch}`
      end
      # clean or create package dir
      if Dir.exist?("pkg\\{@@r_inst_dir}")
        `del /s /q pkg\\{@@r_inst_dir}\\*.*`
      else
        `mkdir pkg\\{@@r_inst_dir}`
      end
      puts
    end

    # Copies spec & mspec repos into ruby repo for testing
    def copy_spec
      #puts "copying rubyspec folder..."
      #`xcopy /s /q /y %REPO_SPEC%   %REPO_RUBY%\\spec\\ruby\\`
      #puts "copying mspec folder..."
      #`xcopy /s /q /y %REPO_MSPEC%  %REPO_RUBY%\\spec\\mspec\\`
      puts "xcopy /s /q /y #{__dir__.gsub(/\//, '\\')}\\spec  #{__dir__.gsub(/\//, '\\')}\\src\ruby\\spec\\"
      `xcopy /s /q /y #{__dir__.gsub(/\//, '\\')}\\spec  #{__dir__.gsub(/\//, '\\')}\\src\ruby\\spec\\`
    end

    # Adds version specific patches to patches array
    def vers_patches
      patches = []
      ary = []
      dirs = Dir.glob("#{__dir__}/patches/*").select { |f| File.directory? f }
      dirs.each { |d|
        next unless /(gte|lt)(\d+)\z/ =~ d
        ary << [ $1, $2.to_i, Dir[ File.join(d, '*.patch') ].reject { |fn| File.basename(fn).start_with?('_') } ]
      }
      ary.each { |e|
        if e[0] == 'lt' && @@r_vers_int < e[1]
          patches.concat(e[2])
        elsif e[0] == 'gte' && @@r_vers_int >= e[1]
          patches.concat(e[2])
        end
      }
      patches
    end

    # Returns an integer equivalent of three digit version string
    # a.b.c => 10000a + 100b + c
    def vers_str_2_int(s)
      t = s[/\A\d+\.\d+\.\d+/]  # parse to d.d.d string
      raise "Invalid version string #{s}" unless t
      t.split('.').map.with_index { |e, i| e.to_i * 100 ** (2 - i) }.sum
    end

  end
end
Prepare.run

# ENV['OPENSSL_102'] = '/e/msys64/var/cache/pacman/pkg/mingw-w64-x86_64-ruby-openssl-1.0.2.l-1-any.pkg.tar.xz'
# ENV['OPENSSL_110'] = '/e/msys64/var/cache/pacman/pkg/mingw-w64-x86_64-ruby-openssl-1.1.0.f-1-any.pkg.tar.xz'
# ENV['R_VERS'] = '2.3.4'

#Prepare.check_openssl